gl: Enables direct rendering to the back buffer

wenjieshen opened the pull request:
Allows the engine to render directly to the screen's back buffer, bypassing intermediate framebuffers when possible.

This change introduces a new `mDirectTarget` flag to indicate whether the rendering target is the back buffer.  When rendering directly, it uses a compose task instead of a blit task and avoids unnecessary framebuffer operations.

This improves rendering performance and reduces memory consumption in scenarios where direct rendering is feasible.

**The WASM binding requires modification to enable this feature.**
```cpp
        attrs.depth = true;
        attrs.stencil = true;
        attrs.antialias = true;
```
Related issues: https://github.com/thorvg/thorvg/issues/3951

### Evaluation

#### Summary
- No regressions in any example.
- **Large animation**: +0.3% overall FPS, +1.2% steady-state FPS.
- **Lottie animations**: +3.4% overall FPS, +4.6% steady-state FPS (with higher jitter).
- **Simple animation (normal resolution)**: +3.0% overall FPS, steady-state is the same.
- **Simple animation (4K)**: +44% overall FPS, +36% steady-state FPS (with slightly lower jitter).
- **WebGL multi-animation test**: After a 1-minute run, both builds stabilize at 60 FPS with similar memory ranges. No meaningful difference observed.

Overall, the feature branch is performance-safe and shows significant improvements at high resolutions, especially in 4K scenarios.

---

### 1. Large Animation
<img width="769" height="785" alt="image" src="https://github.com/user-attachments/assets/76b795a6-a656-4a39-91b1-2510f735e3fd" />

**Overall FPS**  
- Origin: 25.66  
- Feature: 25.74  
- Change: **+0.31%**

**Steady-State FPS (last 20 intervals)**  
- Origin: 25.49  
- Feature: 25.79  
- Change: **+1.18%**

**Jitter (stddev/mean)**  
- Origin: 1.71%  
- Feature: 2.18%  
- Relative change: **+27%**

**Conclusion:** Tiny gain, effectively identical to origin. No regression.

---

### 2. Lottie Animation
<img width="961" height="800" alt="image" src="https://github.com/user-attachments/assets/bfea140d-f0e6-4d6d-832f-acf500eb1539" />

**Overall FPS**  
- Origin: 34.28  
- Feature: 35.45  
- Change: **+3.41%**

**Steady-State FPS**  
- Origin: 34.75  
- Feature: 36.36  
- Change: **+4.62%**

**Jitter**  
- Origin: 3.21%  
- Feature: 5.51%  
- Relative change: **+72%**

**Conclusion:** Clear 3–5% FPS gain with higher variability. Still stable.

---

### 3. Small Animation (Normal Resolution)
<img width="1026" height="1056" alt="image" src="https://github.com/user-attachments/assets/4237e1cf-72c9-401f-9b77-80ea1fc55d28" />

**Overall FPS**  
- Origin: 187.20  
- Feature: 192.72  
- Change: **+2.95%**

**Steady-State FPS**  
- Origin: 239.55  
- Feature: 239.33  
- Change: **−0.09%**

**Jitter**  
- Origin: 0.10%  
- Feature: 0.14%  
- Relative change: **+43%**

**Conclusion:** +3% overall from faster warm-up; steady-state identical.

---

### 4. Small Animation (4K Resolution)
<img width="3841" height="2160" alt="Screenshot 2025-11-23 011229" src="https://github.com/user-attachments/assets/497ea795-dcf9-4c45-acb1-6720043edc8a" />

**Overall FPS**  
- Origin: 120.01  
- Feature: 173.07  
- Change: **+44.21%**

**Steady-State FPS**  
- Origin: 119.04  
- Feature: 161.46  
- Change: **+35.64%**

**Jitter**  
- Origin: 0.85%  
- Feature: 0.67%  
- Change: **−21%**

**Conclusion:** Major performance uplift at 4K — both average and steady-state FPS improve significantly.

---

### 5. Resolution Scaling (Normal → 4K)

#### Origin Branch
- Overall FPS: −35.89%  
- Steady-State FPS: −50.31%

#### Feature Branch
- Overall FPS: −10.20%  
- Steady-State FPS: −32.54%

**Conclusion:** The feature branch is far more resolution-robust.  
Origin loses half its steady-state FPS, while the feature retains much more performance.

---

### 6. WebGL Multi-Animation Test (Qualitative)
<img width="4064" height="2160" alt="Screenshot 2025-11-22 at 8 48 52 PM" src="https://github.com/user-attachments/assets/30f2b311-35ed-490f-90d6-bfda7aabbb51" />

Tested using the WebGL/DevTools HUD for 1 minute.

**Feature Build**
- FPS: 60 (range 29–60)
- Memory counter: ~20 MB (range 19–25 MB)

**Origin Build**
- FPS: 60 (range 29–60)
- Memory counter: ~23 MB (range 19–25 MB)

**Conclusion:**
- FPS stabilizes at 60 on both builds.
- The HUD memory number fluctuates and stays in the same range for both builds.
- Treat this as **no meaningful difference** and, importantly, **no regression**.

**Notes**
- I also make sure that direct rendering is enabled in WebGL with a tool, but the performance remains exactly the same.
<p>
<img width="1016" height="540" alt="image" src="https://github.com/user-attachments/assets/722d3122-d6fb-4308-80dd-eb17822aa571" />
<img width="1016" height="540" alt="image" src="https://github.com/user-attachments/assets/31863db6-7503-4595-8ebe-28dccb1fbfe8" />
</p>

---

tinyjin reviewed (approved):
Looks good!

<img width="1754" height="578" alt="Image" src="https://github.com/user-attachments/assets/9dbccef7-1cb5-4a04-962d-4424dae0b516" />
<img width="1744" height="604" alt="Image" src="https://github.com/user-attachments/assets/04d144a5-ef1b-4176-bc60-d3717d5c3e50" />
<img width="1746" height="576" alt="Image" src="https://github.com/user-attachments/assets/eb117580-c8e9-428c-9a4b-6c31b9eae53c" />
<img width="1740" height="606" alt="Image" src="https://github.com/user-attachments/assets/64e8d25c-892a-4d86-b443-e7b7f3af6f15" />

hermet commented:
@SergeyLebedkin Hello, please confirm the performance beneficial on your side. Thanks.

hermet commented:
> macos/intel Lottie, gl main vs gl <img alt="image" width="424" height="250" src="https://private-user-images.githubusercontent.com/7770034/524221871-3eea08f2-3a50-45fd-99d6-d9de1ac4afa5.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjUyOTE0NDIsIm5iZiI6MTc2NTI5MTE0MiwicGF0aCI6Ii83NzcwMDM0LzUyNDIyMTg3MS0zZWVhMDhmMi0zYTUwLTQ1ZmQtOTlkNi1kOWRlMWFjNGFmYTUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTIwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTEyMDlUMTQzOTAyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODEwMzBhMTY4ZmRmOGRhMTJmN2UwN2Q5NGYxYTU1YzY0YWRlZjk2ZDNlOWZjMWEwMjEwNDhmMTg1MjVlZjM0MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.flpY1CDMe-5QPT3a_JFoAYWrwoz36diXRpZ5KvAb9Bg">

@SergeyLebedkin Could you please test on windows machine? Thanks.

wenjieshen committed:
gl: Enables direct rendering to the back buffer

Allows the engine to render directly to the screen's back buffer,
bypassing intermediate framebuffers when possible.

This change introduces a new `mDirectTarget` flag to indicate whether the rendering target is the back buffer.
When rendering directly, it uses a compose task instead of a blit task, and avoids unnecessary framebuffer operations.

This improves rendering performance and reduces memory consumption in scenarios where direct rendering is feasible.

Related issues: https://github.com/thorvg/thorvg/issues/3951

wenjieshen commented:
> @wenjieshen please rebase & update the `thorvg.example`

Thank you for the reminder. I've rebased this branch and tested it on Windows.

wenjieshen commented:
> @wenjieshen Hello, I tested on Local Linux machine but I **coudln't find any enhancement** at fps in both **Lottie & Simple animation (4K)** with applying [this](https://github.com/thorvg/thorvg.example/pull/6/commits). Any Opinion?

I appreciate your help. Could you please enable the log build option and confirm whether this feature has been enabled? Thank you.

hermet commented:
> > @wenjieshen Hello, I tested on Local Linux machine but I **coudln't find any enhancement** at fps in both **Lottie & Simple animation (4K)** with applying [this](https://github.com/thorvg/thorvg.example/pull/6/commits). Any Opinion?
> 
> I appreciate your help. Could you please enable the log build option and confirm whether this feature has been enabled? Thank you.

@wenjieshen it seems working.
```
 L] GL_ENGINE (../src/renderer/gl_engine/tvgGl.cpp 821): OpenGL/ES version = v4.6
[E] GL_ENGINE (../src/renderer/gl_engine/tvgGlRenderer.cpp 79): Maybe missing MakeCurrent() Call?
[L] GL_ENGINE (../src/renderer/gl_engine/tvgGlRenderer.cpp 956): Default framebuffer stencil/depth size: 8/24
[L] GL_ENGINE (../src/renderer/gl_engine/tvgGlRenderer.cpp 956): Default framebuffer stencil/depth size: 8/24
````

hermet commented:
@wenjieshen + Tested Animation on Windows machine (Ryzen 7 5800X / RTX4700). **No enhancement**.

wenjieshen commented:
> @wenjieshen jfyi, @SergeyLebedkin's test also shows no or lower performance, same as mine.

Yes, I believe direct rendering is enabled on your device. However, I suspect that the default framebuffer path on macOS/Linux may introduce additional overhead or stalls, which could make this pipeline appear slower. To investigate further, I would capture the time spent in the pipeline. That should help us tell whether enabling this introduces extra system overhead that cancels out the time we saved, or whether blitting is truly negligible on these systems.

hermet commented:
@wenjieshen Could you please rebase this one more time? Thanks.

Nor-s commented:
I tested it on 2 machines.

1. Mac (M5pro) - No enhancement 
2. Linux (Arch Linux, AMD 7600, 3060ti)
   1. Lottie - No enhancement 
   2. Animation (1024x1024): 4438.52 -> 5498.4 (+23.88%)
   3. Animation (3440x1440): 2748.385 -> 5974.808 (+117.39%)
   
   
---

The Animation FPS values on the Linux machine were abnormally high compared to the other machine, so I removed a specific optimization before measuring.

```cpp
        //Update animation frame only when it's changed        
        if (animation->frame(animation->totalFrame() * progress) == tvg::Result(0)) {
```

I removed the part that updates conditionally, as shown above, and then ran the test.
The measurements were taken from the example in the link below, and for 3440x1440 I used SDL's fullscreen feature. (SDL_WINDOW_FULLSCREEN_DESKTOP flag)

- [thorvg.example](https://github.com/thorvg/thorvg.example/commit/35ad37795c09ca307a63cf335d99e04647433903#diff-7fbbf0ed412c3437b8254fc0acd4b74818f22abc0ea45ee0defc54799789b317)

hermet commented:
@tinyjin could you please confirm any regression / perf-test on web?

wenjieshen commented:
> @tinyjin could you please confirm any regression / perf-test on web?

Hi @tinyjin,
Thank you for the testing on the web platform.

[Main]
- FPS 53 (52-54)
- 42 (40-45) MB

[PR]
- FPS FPS 56 (55-57) 4.5%+
- 40 (39-41) MB

hermet commented:
> ```c++
>         attrs.depth = true;
>         attrs.stencil = true;
>         attrs.antialias = true;
> ```

@wenjieshen could you please help a PR for this fix together?

hermet commented:
FPS is somewhat fluctuating (< 5fps) and can't confirm any fps enhancement with Lottie (1600x1600) on Intel Arc and Ryzen 7 5800X

hermet commented:
@wenjieshen Let's hold this, we will see again later.
